proposals/p2376.md

```diff
@@ -0,0 +1,201 @@
+this restriction, the set of interfaces known to implement a type would change
+depending on which interfaces definitions are imported and known to be
+satisfied, which is a coherence problem.
+
+This restriction also allows interfaces and named constraints to be used while
+incomplete, which allows some use cases that involve circular references,
+including self reference. The logic goes like this:
+
+-   we want to limit when information from a constraint is found,
+-   to increase the cases where we don't need to look in a constraint,
+-   so constraints are allowed to be incomplete in those cases.
+
+Proposal [#2347](https://github.com/carbon-language/carbon-lang/pull/2347) lists
+conditions when we want to allow constraints to be incomplete.
+
+## Background
+
+There are a number of earlier proposals related to or modified by this proposal:
+
+-   [#553: Generics details part 1](https://github.com/carbon-language/carbon-lang/pull/553)
+    introduced `impl as` restrictions in interfaces and named constraints.
+-   [#818: Constraints for generics (generics details 3)](https://github.com/carbon-language/carbon-lang/pull/818)
+    introduce `where` constraints.
+-   [#1013: Generics: Set associated constants using `where` constraints](https://github.com/carbon-language/carbon-lang/pull/1013)
+    switched to using `where` constraints in `impl` declarations to specify
+    associated constants.
+-   [#1084: Generics details 9: forward declarations](https://github.com/carbon-language/carbon-lang/pull/1084)
+    allowed forward declaration of interfaces and named constraints, explicitly
+    supporting incomplete interfaces and named constraints beyond when they were
+    being defined.
+-   [#2107: Clarify rules around `Self` and `.Self`](https://github.com/carbon-language/carbon-lang/pull/2107)
+    established some rules around `Self` and `.Self`, which this proposal adds
+    to.
+-   [#2347: What can be done with an incomplete interface](https://github.com/carbon-language/carbon-lang/pull/2347)
+    clarifies what can be done with an incomplete interface or named constraint.
+    Those rules rely on this proposal to be implementable.
+
+## Proposal
+
+`where` clauses must use a designator, either `.Self` or `.Foo` for some member
+`Foo`. The designator may be used directly, or supplied as an argument to a
+type, interface, or named constraint used in the `where` clause, as in these
+examples:
+
+-   `Container where .ElementType = i32`
+-   `Type where Vector(.Self) is Sortable`
+-   `Addable where i32 is AddableWith(.Result)`
+
+`impl as` declarations in interfaces and named constraints must always involve
+`Self`:
+
+-   Can be the implicit `Self` when no type is specified, as in `impl as ...`,
+    or the equivalent declarations with `Self` declared explicitly, as in
+    `impl Self as ...`
+-   Can be an argument a type. The type can be what is to the left of the `as`,
```

Comment:
First sentence seems to be missing a word.

Reply:
added "to"